File: backend/src/app/services/shared_service.py
"""Shared Dashboard Service - Business logic for Security & Sharing Dashboard.

This service provides the core business logic for the workspace-wide
Sharing Dashboard, enabling photographers to:
- View all active share links across galleries
- Bulk revoke links (Kill Switch)
- Export sharing data for compliance

Security considerations:
- All operations are workspace-scoped
- Audit logging for revocation actions
- Rate limiting on list operations
"""

from __future__ import annotations

import csv
import io
import json
import logging
from datetime import datetime, timezone
from typing import Any, Optional
from uuid import UUID

from app.repositories.shared_repository import SharedRepository, get_shared_repository


logger = logging.getLogger(__name__)


class SharedServiceError(Exception):
    """Base exception for shared service errors."""

    def __init__(
        self,
        message: str,
        code: str,
        status_code: int = 400,
    ):
        super().__init__(message)
        self.message = message
        self.code = code
        self.status_code = status_code


class LinkNotFoundError(SharedServiceError):
    """Link not found in workspace."""

    def __init__(self, link_id: Optional[str] = None):
        super().__init__(
            message="Link not found" + (f": {link_id}" if link_id else ""),
            code="LINK_NOT_FOUND",
            status_code=404,
        )


class BulkOperationError(SharedServiceError):
    """Error during bulk operation."""

    def __init__(self, message: str):
        super().__init__(
            message=message,
            code="BULK_OPERATION_FAILED",
            status_code=400,
        )


class SharedService:
    """Service for shared dashboard operations.

    Provides business logic for:
    - Listing all magic links across workspace
    - Getting detailed link information
    - Bulk revocation (Kill Switch)
    - Exporting sharing data
    - Aggregate statistics
    """

    # Constants
    MAX_BULK_REVOKE = 100  # Maximum links to revoke in one request
    DEFAULT_STATS_DAYS = 30  # Default period for statistics

    def __init__(self, repository: Optional[SharedRepository] = None):
        self.repo = repository or get_shared_repository()

    # =========================================================================
    # LIST OPERATIONS
    # =========================================================================

    async def list_links(
        self,
        workspace_id: UUID,
        limit: int = 50,
        offset: int = 0,
        status: Optional[str] = None,
        exclude_revoked: bool = False,
        gallery_id: Optional[UUID] = None,
        search: Optional[str] = None,
        sort_by: str = "created_at",
        sort_order: str = "desc",
    ) -> dict[str, Any]:
        """List all magic links across the workspace.

        Args:
            workspace_id: Workspace ID for tenant isolation
            limit: Maximum results (1-100)
            offset: Pagination offset
            status: Filter by status ('active', 'expired', 'revoked', 'all')
            exclude_revoked: If True, exclude revoked links from results
            gallery_id: Filter by specific gallery
            search: Search term for label or gallery title
            sort_by: Column to sort by
            sort_order: Sort direction

        Returns:
            Dict with links array and meta information
        """
        # Validate and sanitize inputs
        limit = min(max(1, limit), 100)
        offset = max(0, offset)

        valid_statuses = {"active", "expired", "revoked", "all", None}
        if status not in valid_statuses:
            status = "active"

        valid_sorts = {"created_at", "access_count", "expires_at", "last_accessed_at"}
        if sort_by not in valid_sorts:
            sort_by = "created_at"

        if sort_order.lower() not in {"asc", "desc"}:
            sort_order = "desc"

        links, counts = await self.repo.list_workspace_links(
            workspace_id=workspace_id,
            limit=limit,
            offset=offset,
            status=status,
            exclude_revoked=exclude_revoked,
            gallery_id=gallery_id,
            search=search,
            sort_by=sort_by,
            sort_order=sort_order,
        )

        return {
            "data": links,
            "meta": {
                "total": counts["total"],
                "active_count": counts["active_count"],
                "expired_count": counts["expired_count"],
                "revoked_count": counts["revoked_count"],
                "limit": limit,
                "offset": offset,
                "has_more": offset + len(links) < counts["total"],
            },
        }

    async def get_link_detail(
        self,
        workspace_id: UUID,
        link_id: UUID,
    ) -> dict[str, Any]:
        """Get detailed information for a single link.

        Args:
            workspace_id: Workspace ID for tenant isolation
            link_id: The magic link ID

        Returns:
            Link details with gallery info and access log

        Raises:
            LinkNotFoundError: If link not found in workspace
        """
        link = await self.repo.get_link_detail(workspace_id, link_id)

        if not link:
            raise LinkNotFoundError(str(link_id))

        return link

    # =========================================================================
    # STATISTICS
    # =========================================================================

    async def get_stats(
        self,
        workspace_id: UUID,
        days: int = 30,
    ) -> dict[str, Any]:
        """Get aggregate sharing statistics for the workspace.

        Args:
            workspace_id: Workspace ID for tenant isolation
            days: Number of days for trend data (1-90)

        Returns:
            Aggregate statistics including totals, trends, breakdowns
        """
        # Validate days
        days = min(max(1, days), 90)

        return await self.repo.get_workspace_stats(workspace_id, days)

    # =========================================================================
    # REVOCATION
    # =========================================================================

    async def revoke_link(
        self,
        workspace_id: UUID,
        link_id: UUID,
        user_id: Optional[UUID] = None,
        reason: Optional[str] = None,
    ) -> dict[str, Any]:
        """Revoke a single magic link.

        Args:
            workspace_id: Workspace ID for tenant isolation
            link_id: The magic link ID
            user_id: User performing the revocation (for audit)
            reason: Optional reason for audit log

        Returns:
            Success message

        Raises:
            LinkNotFoundError: If link not found in workspace
        """
        # Use bulk revoke with single link
        revoked_count, failures = await self.repo.bulk_revoke(
            workspace_id=workspace_id,
            link_ids=[link_id],
        )

        if failures:
            raise LinkNotFoundError(str(link_id))

        logger.info(
            "Magic link revoked via shared dashboard",
            extra={
                "workspace_id": str(workspace_id),
                "link_id": str(link_id),
                "user_id": str(user_id) if user_id else None,
                "reason": reason,
            },
        )

        return {
            "message": "Link revoked successfully",
            "link_id": str(link_id),
        }

    async def bulk_revoke(
        self,
        workspace_id: UUID,
        link_ids: list[UUID],
        user_id: Optional[UUID] = None,
        reason: Optional[str] = None,
    ) -> dict[str, Any]:
        """Revoke multiple magic links at once (Kill Switch).

        Args:
            workspace_id: Workspace ID for tenant isolation
            link_ids: List of link IDs to revoke
            user_id: User performing the revocation (for audit)
            reason: Optional reason for audit log

        Returns:
            Dict with revoked count and any failures

        Raises:
            BulkOperationError: If too many links requested
        """
        if len(link_ids) > self.MAX_BULK_REVOKE:
            raise BulkOperationError(
                f"Cannot revoke more than {self.MAX_BULK_REVOKE} links at once"
            )

        if not link_ids:
            return {
                "revoked_count": 0,
                "failed": [],
            }

        revoked_count, failures = await self.repo.bulk_revoke(
            workspace_id=workspace_id,
            link_ids=link_ids,
        )

        logger.info(
            "Bulk revoke via shared dashboard",
            extra={
                "workspace_id": str(workspace_id),
                "user_id": str(user_id) if user_id else None,
                "requested": len(link_ids),
                "revoked": revoked_count,
                "failed": len(failures),
                "reason": reason,
            },
        )

        return {
            "revoked_count": revoked_count,
            "failed": failures,
        }

    # =========================================================================
    # EXPORT
    # =========================================================================

    async def export_links(
        self,
        workspace_id: UUID,
        format: str = "csv",
        date_from: Optional[datetime] = None,
        date_to: Optional[datetime] = None,
    ) -> tuple[bytes, str, str]:
        """Export sharing data for compliance/audit.

        Args:
            workspace_id: Workspace ID for tenant isolation
            format: Output format ('csv' or 'json')
            date_from: Optional start date filter
            date_to: Optional end date filter

        Returns:
            Tuple of (data bytes, content type, filename)
        """
        links = await self.repo.export_links(
            workspace_id=workspace_id,
            date_from=date_from,
            date_to=date_to,
        )

        timestamp = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")

        if format == "json":
            data = json.dumps(links, indent=2, default=str).encode("utf-8")
            content_type = "application/json"
            filename = f"sharing_export_{timestamp}.json"
        else:
            # CSV format
            output = io.StringIO()
            if links:
                fieldnames = [
                    "link_id",
                    "gallery_id",
                    "gallery_title",
                    "label",
                    "status",
                    "target_type",
                    "expires_at",
                    "max_accesses",
                    "access_count",
                    "created_at",
                    "updated_at",
                    "created_by_email",
                    "password_protected",
                    "pin_protected",
                    "email_registration_required",
                    "download_policy",
                ]
                writer = csv.DictWriter(output, fieldnames=fieldnames)
                writer.writeheader()
                for link in links:
                    writer.writerow({k: link.get(k, "") for k in fieldnames})
            else:
                output.write("No links found for the specified period.\n")

            data = output.getvalue().encode("utf-8")
            content_type = "text/csv"
            filename = f"sharing_export_{timestamp}.csv"

        logger.info(
            "Sharing data exported",
            extra={
                "workspace_id": str(workspace_id),
                "format": format,
                "record_count": len(links),
            },
        )

        return data, content_type, filename


# ---------------------------------------------------------------------------
# Singleton Instance
# ---------------------------------------------------------------------------

_shared_service: Optional[SharedService] = None


def get_shared_service() -> SharedService:
    """Get singleton shared service instance."""
    global _shared_service
    if _shared_service is None:
        _shared_service = SharedService()
    return _shared_service

File: frontend/src/services/sharedService.ts
/**
 * Shared Dashboard Service
 * API client for Security & Sharing Dashboard operations
 *
 * Provides methods for:
 * - Listing all shared links across workspace
 * - Getting aggregate statistics
 * - Bulk revocation (Kill Switch)
 * - Exporting sharing data
 */

import apiClient from './api';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AccessPolicy {
  password_protected: boolean;
  pin_protected: boolean;
  email_required: boolean;
  download_policy: string;
}

export interface RecentVisitor {
  country_code: string | null;
  device_type: string | null;
  accessed_at: string | null;
}

export interface SharedLinkItem {
  link_id: string;
  gallery_id: string;
  gallery_title: string | null;
  gallery_status: string | null;
  label: string | null;
  status: 'active' | 'expired' | 'revoked';
  target_type: 'gallery' | 'sub_gallery' | 'photo';
  target_id: string | null;
  expires_at: string | null;
  max_accesses: number | null;
  access_count: number;
  last_accessed_at: string | null;
  access_policy: AccessPolicy;
  recent_visitors: RecentVisitor[];
  created_at: string;
  created_by_name: string | null;
}

export interface SharedLinksMeta {
  total: number;
  active_count: number;
  expired_count: number;
  revoked_count: number;
  limit: number;
  offset: number;
  has_more: boolean;
}

export interface SharedLinksResponse {
  data: SharedLinkItem[];
  meta: SharedLinksMeta;
}

export interface Visitor {
  visitor_id: string | null;
  email: string | null;
  first_name: string | null;
  last_name: string | null;
}

export interface AccessLogEntry {
  access_id: string;
  accessed_at: string;
  ip_address: string | null;
  country_code: string | null;
  region: string | null;
  city: string | null;
  device_type: string | null;
  browser: string | null;
  os: string | null;
  email_gate_passed: boolean;
  pin_gate_passed: boolean;
  visitor: Visitor | null;
}

export interface SharedLinkDetail extends SharedLinkItem {
  gallery_description: string | null;
  updated_at: string | null;
  created_by_email: string | null;
  accesses: AccessLogEntry[];
}

export interface DeviceBreakdown {
  desktop: number;
  mobile: number;
  tablet: number;
}

export interface DailyAccess {
  date: string;
  count: number;
}

export interface TopGallery {
  gallery_id: string;
  gallery_title: string;
  link_count: number;
  access_count: number;
}

export interface TopCountry {
  country_code: string;
  count: number;
}

export interface SharedStats {
  total_links: number;
  active_links: number;
  expired_links: number;
  revoked_links: number;
  total_accesses_period: number;
  unique_visitors_period: number;
  period_days: number;
  device_breakdown: DeviceBreakdown;
  accesses_by_day: DailyAccess[];
  top_galleries: TopGallery[];
  top_countries: TopCountry[];
}

export interface BulkRevokeFailure {
  link_id: string;
  error: string;
}

export interface BulkRevokeResponse {
  revoked_count: number;
  failed: BulkRevokeFailure[];
}

export interface ListLinksOptions {
  limit?: number;
  offset?: number;
  status?: 'active' | 'expired' | 'revoked' | 'all';
  exclude_revoked?: boolean; // Hide revoked links from list (default: false for backward compatibility)
  gallery_id?: string;
  search?: string;
  sort_by?: 'created_at' | 'access_count' | 'expires_at' | 'last_accessed_at';
  sort_order?: 'asc' | 'desc';
}

// ---------------------------------------------------------------------------
// Error Handling
// ---------------------------------------------------------------------------

export class SharedServiceError extends Error {
  code: string;
  status: number;

  constructor(message: string, code: string, status: number) {
    super(message);
    this.name = 'SharedServiceError';
    this.code = code;
    this.status = status;
  }
}

const ERROR_MESSAGES: Record<string, string> = {
  LINK_NOT_FOUND: 'This link could not be found',
  BULK_OPERATION_FAILED: 'Failed to revoke some links',
  INTERNAL_ERROR: 'Something went wrong. Please try again.',
};

function getErrorMessage(code: string, fallback?: string): string {
  return ERROR_MESSAGES[code] || fallback || 'An error occurred';
}

// ---------------------------------------------------------------------------
// Service Class
// ---------------------------------------------------------------------------

export class SharedService {
  /**
   * List all shared links across the workspace
   */
  async listLinks(
    workspaceId: string,
    options?: ListLinksOptions
  ): Promise<SharedLinksResponse> {
    const params = new URLSearchParams();

    if (options?.limit) params.append('limit', options.limit.toString());
    if (options?.offset) params.append('offset', options.offset.toString());
    if (options?.status) params.append('status', options.status);
    if (options?.exclude_revoked !== undefined) params.append('exclude_revoked', options.exclude_revoked.toString());
    if (options?.gallery_id) params.append('gallery_id', options.gallery_id);
    if (options?.search) params.append('search', options.search);
    if (options?.sort_by) params.append('sort_by', options.sort_by);
    if (options?.sort_order) params.append('sort_order', options.sort_order);

    const query = params.toString();
    const endpoint = `/api/v1/workspaces/${workspaceId}/shared/links${query ? `?${query}` : ''}`;

    const response = await apiClient.get<SharedLinksResponse>(endpoint);
    if (response.error) {
      throw new SharedServiceError(
        getErrorMessage(response.error.code, response.error.message),
        response.error.code || 'INTERNAL_ERROR',
        response.error.status || 500
      );
    }
    return response.data!;
  }

  /**
   * Get detailed information for a single link
   */
  async getLinkDetail(
    workspaceId: string,
    linkId: string
  ): Promise<SharedLinkDetail> {
    const endpoint = `/api/v1/workspaces/${workspaceId}/shared/links/${linkId}`;

    const response = await apiClient.get<SharedLinkDetail>(endpoint);
    if (response.error) {
      throw new SharedServiceError(
        getErrorMessage(response.error.code, response.error.message),
        response.error.code || 'INTERNAL_ERROR',
        response.error.status || 500
      );
    }
    return response.data!;
  }

  /**
   * Get aggregate sharing statistics
   */
  async getStats(workspaceId: string, days: number = 30): Promise<SharedStats> {
    const endpoint = `/api/v1/workspaces/${workspaceId}/shared/stats?days=${days}`;

    const response = await apiClient.get<SharedStats>(endpoint);
    if (response.error) {
      throw new SharedServiceError(
        getErrorMessage(response.error.code, response.error.message),
        response.error.code || 'INTERNAL_ERROR',
        response.error.status || 500
      );
    }
    return response.data!;
  }

  /**
   * Revoke a single link
   */
  async revokeLink(workspaceId: string, linkId: string): Promise<void> {
    const endpoint = `/api/v1/workspaces/${workspaceId}/shared/links/${linkId}`;

    const response = await apiClient.delete<{ message: string }>(endpoint);
    if (response.error) {
      throw new SharedServiceError(
        getErrorMessage(response.error.code, response.error.message),
        response.error.code || 'INTERNAL_ERROR',
        response.error.status || 500
      );
    }
  }

  /**
   * Bulk revoke multiple links (Kill Switch)
   */
  async bulkRevoke(
    workspaceId: string,
    linkIds: string[],
    reason?: string
  ): Promise<BulkRevokeResponse> {
    const endpoint = `/api/v1/workspaces/${workspaceId}/shared/bulk-revoke`;

    const response = await apiClient.post<BulkRevokeResponse>(endpoint, {
      link_ids: linkIds,
      reason,
    });
    if (response.error) {
      throw new SharedServiceError(
        getErrorMessage(response.error.code, response.error.message),
        response.error.code || 'INTERNAL_ERROR',
        response.error.status || 500
      );
    }
    return response.data!;
  }

  /**
   * Export sharing data for compliance
   */
  async exportData(
    workspaceId: string,
    options?: {
      format?: 'csv' | 'json';
      date_from?: string;
      date_to?: string;
    }
  ): Promise<Blob> {
    const params = new URLSearchParams();
    if (options?.format) params.append('format', options.format);
    if (options?.date_from) params.append('date_from', options.date_from);
    if (options?.date_to) params.append('date_to', options.date_to);

    const query = params.toString();
    const endpoint = `/api/v1/workspaces/${workspaceId}/shared/export${query ? `?${query}` : ''}`;

    const response = await fetch(endpoint, {
      method: 'GET',
      credentials: 'include',
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new SharedServiceError(
        getErrorMessage(error.code, error.message),
        error.code || 'INTERNAL_ERROR',
        response.status
      );
    }

    return response.blob();
  }

  /**
   * Download export file
   */
  async downloadExport(
    workspaceId: string,
    options?: {
      format?: 'csv' | 'json';
      date_from?: string;
      date_to?: string;
    }
  ): Promise<void> {
    const blob = await this.exportData(workspaceId, options);
    const format = options?.format || 'csv';
    const filename = `sharing_export_${new Date().toISOString().slice(0, 10)}.${format}`;

    const url = URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(url);
  }

  // ---------------------------------------------------------------------------
  // Utility Methods
  // ---------------------------------------------------------------------------

  /**
   * Format expiration for display
   */
  formatExpiration(expiresAt: string | null | undefined): string {
    if (!expiresAt) return 'Never expires';

    const expDate = new Date(expiresAt);
    const now = new Date();

    if (expDate < now) {
      return 'Expired';
    }

    const diff = expDate.getTime() - now.getTime();
    const days = Math.floor(diff / (1000 * 60 * 60 * 24));
    const hours = Math.floor((diff % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));

    if (days > 30) {
      return `Expires ${expDate.toLocaleDateString()}`;
    } else if (days > 0) {
      return `${days}d remaining`;
    } else if (hours > 0) {
      return `${hours}h remaining`;
    } else {
      return 'Expires soon';
    }
  }

  /**
   * Format access count for display
   */
  formatAccessCount(count: number, max?: number | null): string {
    if (max) {
      return `${count} / ${max}`;
    }
    return count.toString();
  }

  /**
   * Get status badge color
   */
  getStatusColor(status: string): string {
    switch (status) {
      case 'active':
        return 'success';
      case 'expired':
        return 'warning';
      case 'revoked':
        return 'error';
      default:
        return 'default';
    }
  }

  /**
   * Get country name from code
   */
  getCountryName(code: string): string {
    const countries: Record<string, string> = {
      US: 'United States',
      GB: 'United Kingdom',
      CA: 'Canada',
      AU: 'Australia',
      DE: 'Germany',
      FR: 'France',
      IN: 'India',
      JP: 'Japan',
      BR: 'Brazil',
      XX: 'Unknown',
    };
    return countries[code] || code;
  }

  /**
   * Get device icon name
   */
  getDeviceIcon(deviceType: string | null): string {
    switch (deviceType) {
      case 'mobile':
        return 'Smartphone';
      case 'tablet':
        return 'Tablet';
      case 'desktop':
      default:
        return 'Monitor';
    }
  }
}

// Export singleton instance
export const sharedService = new SharedService();
export default sharedService;
